4.0 exception hierarchy

martin-neotech opened the pull request:
Clean up exception hierarchy

The exceptions/errors is now grouped and well divided.



Driver API Errors
=================

**neo4j.exceptions.py**

+ Neo4jError
  + ClientError
    + CypherSyntaxError
    + CypherTypeError
    + ConstraintError
    + AuthError
    + Forbidden
      + ForbiddenOnReadOnlyDatabaseError
    + NotALeader
  + DatabaseError
  + TransientError
    + DatabaseUnavailable

+ DriverError
  + TransactionError
  + SessionExpired
  + ServiceUnavailable
     + RoutingServiceUnavailable
     + WriteServiceUnavailable
     + ReadServiceUnavailable
  + ConfigurationError
    + AuthConfigurationError
    + CertificateConfigurationError


Connector API Errors
====================

**neo4j._exceptions.py**

+ BoltError
  + BoltHandshakeError
  + BoltRoutingError
  + BoltConnectionError
  + BoltFailure
  + BoltNeo4jAvailabilityError
  + BoltSecurityError
  + BoltIncompleteCommitError
  + BoltProtocolError
  + Bolt*


```
client_errors = {
    # ConstraintError
    "Neo.ClientError.Schema.ConstraintValidationFailed": ConstraintError,
    "Neo.ClientError.Schema.ConstraintViolation": ConstraintError,
    "Neo.ClientError.Statement.ConstraintVerificationFailed": ConstraintError,
    "Neo.ClientError.Statement.ConstraintViolation": ConstraintError,

    # CypherSyntaxError
    "Neo.ClientError.Statement.InvalidSyntax": CypherSyntaxError,
    "Neo.ClientError.Statement.SyntaxError": CypherSyntaxError,

    # CypherTypeError
    "Neo.ClientError.Procedure.TypeError": CypherTypeError,
    "Neo.ClientError.Statement.InvalidType": CypherTypeError,
    "Neo.ClientError.Statement.TypeError": CypherTypeError,

    # Forbidden
    "Neo.ClientError.General.ForbiddenOnReadOnlyDatabase": ForbiddenOnReadOnlyDatabase,
    "Neo.ClientError.General.ReadOnly": Forbidden,
    "Neo.ClientError.Schema.ForbiddenOnConstraintIndex": Forbidden,
    "Neo.ClientError.Schema.IndexBelongsToConstraint": Forbidden,
    "Neo.ClientError.Security.Forbidden": Forbidden,
    "Neo.ClientError.Transaction.ForbiddenDueToTransactionType": Forbidden,

    # AuthError
    "Neo.ClientError.Security.AuthorizationFailed": AuthError,
    "Neo.ClientError.Security.Unauthorized": AuthError,

    # NotALeader
    "Neo.ClientError.Cluster.NotALeader": NotALeader
}

transient_errors = {

    # DatabaseUnavailableError
    "Neo.TransientError.General.DatabaseUnavailable": DatabaseUnavailable
}
```

martin-neotech committed:
Changed CypherError to Neo4jError

martin-neotech committed:
Removed unused error.

martin-neotech committed:
Neo4jAvailabilityError renamed to BoltNeo4jAvailabilityError

martin-neotech committed:
Removed NotALeader and use NotALeaderError

martin-neotech committed:
Removed ForbiddenOnReadOnlyDatabase

Using ForbiddenOnReadOnlyDatabaseError instead

martin-neotech committed:
Divided Driver API Errors and Connector API Errors

The Driver API Errors can be found in exceptions.py
The Connector API Errors can be found in errors.py

martin-neotech committed:
Added DriverError Group

SessionExpired(DriverError)
TransactionError(DriverError)
ServiceUnavailable(DriverError)

martin-neotech committed:
SecurityError is replaced by BoltSecurityError

Driver API ClientErrors does not inherit SecurityError and SecurityError
is removed from the Driver API Errors.

martin-neotech committed:
IncompleteCommitError is replaced with BoltIncompleteCommitError

martin-neotech committed:
ProtocolError is replaced with BoltProtocolError

technige commented on neo4j/errors.py:
```diff
@@ -168,53 +168,20 @@ def transaction(self):
-
-
-class Neo4jAvailabilityError(Neo4jError):
+class BoltNeo4jAvailabilityError(BoltError):
```
This is a very clumsy name. It should have either one prefix or the other, not both. How come this has moved from surface (`Neo4jError`) to internal (`BoltError`)?

martin-neotech replied:
`BoltNeo4jAvailabilityError` seems to be a mixture of `ServiceUnavailable` errors. That is why I categorized it as an internal error. I think its best to add sub-categories for each unavailable error.

Will refactor the errors to be under `ServiceUnavailable`

`RoutingServiceUnavailable(ServiceUnavailable)`
"Unable to retrieve routing information"

`WriteServiceUnavailable(ServiceUnavailable)`
"No write service available"

`ReadServiceUnavailable(ServiceUnavailable)`
"No read service available"

martin-neotech replied:
Removed `BoltNeo4jAvailabilityError`  and replaced the ones with sub classes for `ServiceUnavailable`

+ ServiceUnavailable
     + RoutingServiceUnavailable
     + WriteServiceUnavailable
     + ReadServiceUnavailable

Commit: https://github.com/neo4j/neo4j-python-driver/pull/350/commits/06279ad6946e23741a680f6c300b119f5bcf997f

technige commented on docs/source/errors.rst:
```diff
@@ -11,33 +11,19 @@ Connectivity errors
-
-    Raised when an unexpected or unsupported protocol event occurs.
-    This error generally indicates a fault with the driver or server software.
-    If you receive this error, please raise a GitHub issue or a support ticket.
```
It's important to leave a note in here as a catch-all. If users see an internal error, in particular a protocol error, they should open an issue.

martin-neotech replied:
Updated the docs about internal errors.

Commit: https://github.com/neo4j/neo4j-python-driver/pull/350/commits/c60e103c810af19963d97fe9b8036c60540d5ab7

I will focus on the docs later on.

technige commented on docs/source/errors.rst:
```diff
@@ -11,33 +11,19 @@ Connectivity errors
 =======================
 
-.. class:: neo4j.exceptions.CypherError
+.. class:: neo4j.exceptions.Neo4jError
```
We need to add this change to the release notes, as it's one that will require users to change their code.

technige reviewed (commented):
While it doesn't have to be part of this PR, we should either collapse `errors` and `exceptions` back into a single module before release, or rename to `exceptions` and `_exceptions`.

martin-neotech committed:
Removed BoltNeo4jAvailabilityError

Added ServiceUnavailable errors that replace BoltNeo4jAvailabilityError
errors.

from neo4j.exceptions import (
    RoutingServiceUnavailable,
    ReadServiceUnavailable,
    WriteServiceUnavailable,
)

martin-neotech committed:
Added docs info about internal driver errors

It's important to leave a note in here as a catch-all.
If users see an internal error, in particular a protocol error, they should open an issue.

martin-neotech committed:
Renamed errors.py to _exceptions.py

The BoltError are internal errors that the user should never experience.

martin-neotech committed:
Changed NotALeaderError to NotALeader

To align with the naming of the status code

Neo.ClientError.Cluster.NotALeader

martin-neotech committed:
Renamed DatabaseUnavailableError to DatabaseUnavailable

This is to align with the naming of status code:

Neo.TransientError.General.DatabaseUnavailable

martin-neotech committed:
Renamed ForbiddenOnReadOnlyDatabaseError to ForbiddenOnReadOnlyDatabase

This is to align with the status code:

Neo.ClientError.General.ForbiddenOnReadOnlyDatabase
